MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c

```diff
@@ -13,30 +13,119 @@ MU_CHANGE: new file
 #include <Library/DeviceStateLib.h>
 #include <Library/DebugLib.h>
 
+CHAR8  *DEVICE_STATE_STRINGS[6] = {
+  DEVICE_STATE_SECUREBOOT_OFF_STRING,
+  DEVICE_STATE_MANUFACTURING_MODE_STRING,
+  DEVICE_STATE_DEVELOPMENT_BUILD_ENABLED_STRING,
+  DEVICE_STATE_SOURCE_DEBUG_ENABLED_STRING,
+  DEVICE_STATE_UNDEFINED_STRING,
+  DEVICE_STATE_UNIT_TEST_MODE_STRING,
+};
+
 /**
-Function to Get current device state
-@retval the current DEVICE_STATE
-**/
-DEVICE_STATE
+ * Get String that represents the device's current insecure states, using the platform defined set of
+ * insecure device states. This can be used to measure insecure device states into the TPM or perform
+ * other required platform actions when the device enters an insecure state.
+ *
+ * @param[out] Buffer - Buffer to store the string
+ * @param[in] MaxSize - Maximum size of the buffer
+ * @retval The The number of characters in the string, including the null character.
+ */
+UINTN
 EFIAPI
-GetDeviceState (
+GetInsecureDeviceStateString (
+  OUT CHAR8  *Buffer,
+  IN UINTN   MaxSize
   )
 {
-  DEVICE_STATE  DevState = PcdGet32 (PcdDeviceStateBitmask);
+  CHAR8         *String;
+  DEVICE_STATE  BitmaskFromIndex;
+  DEVICE_STATE  CurrentDeviceState         = GetDeviceState ();
+  DEVICE_STATE  InsecureDeviceStateSetting = GetInsecureDeviceStateSetting ();
+
+  UINT32  NumberOfDeviceStates = sizeof (DEVICE_STATE_STRINGS) / sizeof (DEVICE_STATE_STRINGS[0]);

```

Comment:
If I understand correctly, I suggest calling this variable `NumberOfDeviceStateStrings`.

```suggestion
  UINT32  NumberOfDeviceStrings = sizeof (DEVICE_STATE_STRINGS) / sizeof (DEVICE_STATE_STRINGS[0]);
```